### PR TITLE
[Internal][Executor] Add tool related apis for long running server

### PR DESCRIPTION
**Pull request by PeiwenGaoMS**
# Description

## App Launch
`uvicorn promptflow.executor._service.app:app --host 0.0.0.0 --port 8000 --reload`

## Currently Ready APIs (Long running server)
**execution.py**
  - /execution/flow: for flow test
  - /execution/node: for single node run

**tool.py**
  - /tool/package_tools: list package tools
  - /tool/meta: generate tool meta

**common.py**
  - /health: get server status
  - /version: get executor version and feature list

## Major Changes
This pull request primarily focuses on refining error handling, enhancing logging, and expanding functionality related to tool metadata generation in the `promptflow` package. The changes include the introduction of new exception classes, the modification of logging context functions, and the addition of a new API endpoint for tool metadata generation.

Error Handling Enhancements:
* [`src/promptflow/promptflow/_core/_errors.py`](diffhunk://#diff-d7c75a55b795a9cdff3801c0911a2cbd102961d8c1f7031aaf36169e1701f9c9R1-R4): Introduced `GenerateMetaSystemError` as a base system exception for tool validation failures and `GenerateMetaTimeout` for handling tool metadata generation timeouts. Updated the docstring of `GenerateMetaUserError` for better clarity. [[1]](diffhunk://#diff-d7c75a55b795a9cdff3801c0911a2cbd102961d8c1f7031aaf36169e1701f9c9R1-R4) [[2]](diffhunk://#diff-d7c75a55b795a9cdff3801c0911a2cbd102961d8c1f7031aaf36169e1701f9c9L128-R132) [[3]](diffhunk://#diff-d7c75a55b795a9cdff3801c0911a2cbd102961d8c1f7031aaf36169e1701f9c9R146-R156) [[4]](diffhunk://#diff-f89131a27c96f14dde04bfaa5d055220eb3b966f6dc40312bb09730e3b539f8bR21-R29)

Logging Improvements:
* [`src/promptflow/promptflow/executor/_service/utils/service_utils.py`](diffhunk://#diff-7ae86db64b00917df8e753fa4577aff711e6494b56658f2f1ba3675257516b23L17-R23): Consolidated `get_log_context` and `get_service_log_context` into a single function with an additional `enable_service_logger` parameter.
* [`src/promptflow/promptflow/executor/_service/apis/execution.py`](diffhunk://#diff-ec7f12cd535f012f940598cdcaa3bc00dccd1cb9d09303b88a50895491a1af17L25-R24): Updated the logging context in `flow_execution` and `node_execution` to use the consolidated `get_log_context` function. [[1]](diffhunk://#diff-ec7f12cd535f012f940598cdcaa3bc00dccd1cb9d09303b88a50895491a1af17L25-R24) [[2]](diffhunk://#diff-ec7f12cd535f012f940598cdcaa3bc00dccd1cb9d09303b88a50895491a1af17L47-R46)

Tool Metadata Generation:
* [`src/promptflow/promptflow/executor/_service/apis/tool.py`](diffhunk://#diff-6ce9d9d4010efc2f2e4308453e14384edaa7f3a5abc8098e1fee97423cb7d6cdR1-R67): Added a new API endpoint for generating tool metadata. This includes handling for tool metadata generation timeouts.
* [`src/promptflow/promptflow/executor/_service/app.py`](diffhunk://#diff-e6bfdc3fdaada257b87c7e783dad24a5757b211b7a9231ac7e26bbe841133289R10-R17): Included the new tool API router in the FastAPI application.
* [`src/promptflow/promptflow/executor/_service/contracts/tool_request.py`](diffhunk://#diff-27b463fa119875a197728e83d46e46e298f7dcaca186e2f1889d88f67d33bd4fR1-R24): Added new request models `ToolMetaRequest` and `RetrieveToolFuncResultRequest` for tool metadata generation and function result retrieval respectively.

Minor Changes:
* [`src/promptflow/promptflow/executor/_service/utils/process_utils.py`](diffhunk://#diff-7984565539b67d71a624abd23cbc7681381333d5935898ace86bc27cc19dbd43L37-R37): Renamed `execute_target_function` to `_execute_target_function` to indicate its private usage. [[1]](diffhunk://#diff-7984565539b67d71a624abd23cbc7681381333d5935898ace86bc27cc19dbd43L37-R37) [[2]](diffhunk://#diff-7984565539b67d71a624abd23cbc7681381333d5935898ace86bc27cc19dbd43L71-R71)
* [`src/promptflow/promptflow/executor/_service/contracts/execution_request.py`](diffhunk://#diff-dd9d37fb5742cebf79e208aae0a7a2defad253ab4d5020eb2c6f609cdf9a3dffL20-R20): Made `log_path` an optional field in `BaseExecutionRequest`.
* `src/promptflow/promptflow/_core/_errors.py` and `src/promptflow/promptflow/executor/_service/apis/tool.py`: Added copyright headers. [[1]](diffhunk://#diff-d7c75a55b795a9cdff3801c0911a2cbd102961d8c1f7031aaf36169e1701f9c9R1-R4) [[2]](diffhunk://#diff-6ce9d9d4010efc2f2e4308453e14384edaa7f3a5abc8098e1fee97423cb7d6cdR1-R67)

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/executor/_service/apis/tool.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        func_input_params_dict=request.func_kwargs,
+        ws_triple_dict=request.ws_triple,
+    )
+
+
+@router.post("/meta")
+async def gen_tool_meta(request: ToolMetaRequest):
+    result = {"tools": {}, "errors": {}}
+    try:
+        result = await invoke_sync_function_in_process(gen_meta, args=(request,), wait_timeout=SHORT_WAIT_TIMEOUT)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_service/apis/tool.py:41]: invoke sync function but the function is actually async?

File: src/promptflow/promptflow/executor/_service/apis/tool.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            if source not in tool_dict and source not in error_dict:
+                error_dict[source] = generate_error_response(GenerateMetaTimeout(source)).to_dict()
+    return result
+
+
+async def gen_meta(request: ToolMetaRequest):
+    with _change_working_dir(request.working_dir), inject_sys_path(request.working_dir):
+        tool_dict = {}
+        error_dict = {}
+        for source, config in request.tools.items():
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_service/apis/tool.py:56]: I remember we already have a function in pf codebase?

File: src/promptflow/promptflow/executor/_service/apis/tool.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+@router.get("/package_tools")
+async def list_package_tools():
+    return collect_package_tools()
+
+
+@router.post("/retrieve_tool_func_result")
+async def retrieve_tool_func_result(request: RetrieveToolFuncResultRequest):
+    from promptflow._core.tools_manager import retrieve_tool_func_result
+
+    return retrieve_tool_func_result(
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_service/apis/tool.py:29]: Not in subprocess?

File: src/promptflow/promptflow/executor/_service/apis/tool.py (hunk 1 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_service/apis/tool.py:41]: Updated.

File: src/promptflow/promptflow/executor/_service/apis/tool.py (hunk 3 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_service/apis/tool.py:29]: Updated.

File: src/promptflow/promptflow/executor/_service/apis/tool.py (hunk 2 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_service/apis/tool.py:52]: Offline synced. We will extract a common function to tool_meta_generator and sdk can reuse it.

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 4]
```diff
@@ -356,6 +372,60 @@ def generate_tool_meta_dict_by_file(path: str, tool_type: ToolType):
...
+                        message_format="Tool type not defined for source '{source}'.",
+                        source=source,
+                    )
+                tool_type = ToolType(config.get("tool_type"))
+                tool_dict[source] = generate_tool_meta_dict_by_file(source, tool_type)
+            except Exception as e:
+                exception_dict[source] = ExceptionPresenter.create(e).to_dict()
+
+
+def generate_tool_meta_in_process(
```
Comment (elliotzh) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:395]: in subprocess?

**Comment by elliotzh** [comment]
@PeiwenGaoMS could you please share me some background about why we need this service? and what level is it in? I mean, we already have a local PFS for control plane and language specified execution service for language specific execution and meta data generation. What is the relationship between this service and the above 2?

**Comment by PeiwenGaoMS** [comment]
@elliotzh In the future, the runtime and executor will run in two separate containers to achieve environmental isolation. Line results will be obtained via API calls instead of function calls.

Therefore, this service is similar to SDK, both calling executor to get the result, and many methods can be shared. so, we hope to extract some common functions to simplify future maintenance.

As for the contract, my understanding is that it should remain consistent as much as possible. Unless there are exceptional circumstances, there should not be a need to maintain two sets of contracts.

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 4 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/_core/tool_meta_generator.py:395]: Updated.

**Review by elliotzh — APPROVED** [review]

**Review by thy09 — DISMISSED** [review]

**Review by elliotzh — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**